### PR TITLE
mod: zora create

### PR DESCRIPTION
**Pull request by stephancill**
The Zora Create mini app allows users to easily and instantly upload an image and turn it into an NFT.

Besides the miniapp itself, it also introduces a change to `@modprotocol/core` which adds base 64 encoded file data to the file picker return value.

The Zora create API requires the `ZORA_ADMIN_PRIVATE_KEY` account, which will be creating and owning the NFT contracts to be registered with an email address on the zora website. The user's ethereum address is automatically set as the royalties recipient.

### CODE REVIEW

File: examples/api/src/app/api/zora-create/route.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import { NextRequest, NextResponse } from "next/server";
+import { PremintAPI, PremintResponse } from "@zoralabs/premint-sdk";
+import { zora } from "viem/chains";
+import { createWalletClient, http } from "viem";
+import { privateKeyToAccount } from "viem/accounts";
+import { NFTStorage } from "nft.storage";
+
+const { NFT_STORAGE_API_KEY } = process.env;
```
Comment (davidfurlong) [inline examples/api/src/app/api/zora-create/route.ts:8]: needs to be added to the `.env.sample`

File: examples/api/src/app/api/zora-create/route.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+  // Upload to NFT.storage // TODO: Detect mime type
+  const [dataPrefix, b64data] = imageData.split(",");
+  const mimeType = dataPrefix.split(";")[0].split(":")[1];
+  const imageBlob = new Blob([Buffer.from(b64data, "base64")], {
+    type: mimeType,
+  });
+
+  const timeString = new Date().toISOString();
+  const title = `Zora Create Mod - ${creator.id} - ${timeString}`;
```
Comment (davidfurlong) [inline examples/api/src/app/api/zora-create/route.ts:59]: I think we should let people set a title

File: examples/api/src/app/api/zora-create/route.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    image: imageBlob,
+    name: title,
+    description: `This digital collectible was created using the Zora Create Mod.`,
+  });
+
+  // Upload to zora
+  const premintAPI = new PremintAPI(zora);
+
+  const adminAccount = privateKeyToAccount(
+    process.env.ZORA_ADMIN_PRIVATE_KEY as `0x${string}`
```
Comment (davidfurlong) [inline examples/api/src/app/api/zora-create/route.ts:71]: add to .env.sample

**Review by davidfurlong — CHANGES_REQUESTED** [review]
looks good, a couple small things remain

File: mods/zora-create/package.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "@miniapps/zora-create",
```
Comment (davidfurlong) [inline mods/zora-create/package.json:2]: @mods/

**Review by davidfurlong — CHANGES_REQUESTED** [review]

**Comment by davidfurlong** [comment]
excited to try this!

File: examples/api/src/app/api/zora-create/route.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+  // Upload to NFT.storage // TODO: Detect mime type
+  const [dataPrefix, b64data] = imageData.split(",");
+  const mimeType = dataPrefix.split(";")[0].split(":")[1];
+  const imageBlob = new Blob([Buffer.from(b64data, "base64")], {
+    type: mimeType,
+  });
+
+  const timeString = new Date().toISOString();
+  const title = prefilledTitle || `Zora Create Mod - ${timeString}`;
```
[the remark this answers is not in the document]
Comment (stephancill) [reply examples/api/src/app/api/zora-create/route.ts:62]: What do you mean by "a cast.fun default title"?

How about `Untitled 12/12/2023` using `new Date().localeDateString`?

Onmod is a bit awkward imo

[the remark this answers is not in the document]
Comment (davidfurlong) [reply examples/api/src/app/api/zora-create/route.ts]: As in "Cast.fun #{timestamp}" or something. Onmod is awk I agree, so is that date format. not sure

[the remark this answers is not in the document]
Comment (stephancill) [reply examples/api/src/app/api/zora-create/route.ts]: Do you have an alternative date format in mind?

[the remark this answers is not in the document]
Comment (davidfurlong) [reply examples/api/src/app/api/zora-create/route.ts]: `31st March, 2021`

[the remark this answers is not in the document]
Comment (stephancill) [reply examples/api/src/app/api/zora-create/route.ts]: added

**Comment by davidfurlong** [comment]
> The Zora create API requires the ZORA_ADMIN_PRIVATE_KEY account, which will be creating and owning the NFT contracts to be registered with an email address on the zora website. The user's ethereum address is automatically set as the royalties recipient.

This introduces a rug risk right? Integrators that host their own mod api can use their own admin key, then at some point change the royalty recipient or other settings?

**Review by davidfurlong — APPROVED** [review]
lgtm besides the package version

**Comment by stephancill** [comment]
> This introduces a rug risk right? Integrators that host their own mod api can use their own admin key, then at some point change the royalty recipient or other settings?

It's a trade off that was made to allow for a wallet-less experience. 

Another option is to make use of message signing and leave it up to the client whether they want to abstract it away or not depending on if they have access to the user's keys or not. I think that may actually be a better way of going about it.

**Comment by davidfurlong** [comment]
happy to do it the current way for now - but we should make a note of this limitation somewhere, perhaps in the readme of the mod or something

**Comment by stephancill** [comment]
Upgraded to the latest version of the zora sdk, had to create a patch a bug which was using the wrong account to sign the contract creation message. If it looks good to you we should be good to merge

**Review by davidfurlong — APPROVED** [review]

**Comment by stephancill** [comment]
Tracking patched issue here

https://github.com/ourzora/zora-protocol/issues/398

**Pull request merged**